proposals/0033-data-collections.md

```diff
@@ -0,0 +1,244 @@
+
+## Implementation
+
+Data collections should following the API design of content collections with a stripped-down featureset. To wire up data collections, we will introduce an internal utility that mirrors our `addContentEntryType()` integration function. This example registers a new `dataEntryType` for `.json` files, with necessary logic to parse data as a JS object:
+
+```ts
+addDataEntryType({
+  extensions: ['.json'],
+  getEntryInfo({ contents, fileUrl }) {
+    // Handle empty JSON files, which cause `JSON.parse` to throw
+    if (contents === undefined || contents === '') return { data: {} };
+
+    const data = JSON.parse(contents);
+
+    if (data == null || typeof data !== 'object')
+      throw new Error(`JSON collection entry ${fileUrl.pathname} must be an object.`);
+
+    return { data };
+  },
+});
+```
+
+Then, we will update our type generator to recognize these data-specific file extensions. This should also raise errors when collections are misconfigured (i.e. `type: 'data'` is missing from the config file) and when a mix of content and data in the same collection is detected.
+
+The `astro:content` runtime module should also be updated to glob these file extensions, and respect the entry ID from the new `getEntry()` utility function. 
+
+# Testing Strategy
+
+- Fixture tests defining data collections, ensuring schemas are respected and query APIs (`getCollection()` and `getEntry()`) return entries of the correct type.
+- Test error states: mixed data / content collections are not allowed, `type: 'data'` is enforced
+
+# Drawbacks
+
+- JSON could be considered added complexity, when users can create `.md` files storing all data via frontmatter instead. Though true for trivial content, this is restrictive for use cases like i18n lookup files (which are typically JSON).
+
+# Alternatives
+
+## Using multi-file vs. single-file
+
+Early proposals supported single0file data collections. This allows storing _all_ data collection entries as an array in one `.json` file, instead of splitting up entries per-file as we do content collections today. For example, a single `src/content/authors.json` file instead of a few `src/content/authors/[name].json` files.
+
+We want to stick with a single API design for an experimental release. Unlike multi-file, there are some prohibitive reasons against single-file collections: 
+
+- **Big arrays don't scale for complex data entries**, like i18n translation files. Users will likely want to split up by file here, especially where the status quo is `en.json | fr.json | jp.json ...` for this use case.
+- **We'd need to parse the whole array of entries** to determine entry IDs. This could be a performance bottleneck vs. pulling IDs from file names.
+- **It would be different from content collections,** which means a learning curve.
+
+Due to these, we decided against single-file for now. Though we do recognize the convenience of colocation that can be explored in the future.
```

Comment:
+1 to exploring in the future. A solution that I'd love to see explored more in a future RFC that I think would kill a few birds with one stone:

```diff
+ import authorsData from '../authors.json';

const authors = defineCollection({
  type: 'data',
  schema: z.object({
    name: z.string(),
    twitter: z.string().url(),
  }),
+  data: authorsData,
});
```

Just a thought / not at all blocking this RFC which I agree is smart to treat this as out-of-scope!

Reply:
What's the use-case for importing like this? It won't work with remote data, I'm surprised at the suggestion.